### ISSUE TITLE
impossible to have the dot / period / decimal on a french keyboard

### ISSUE THREAD

**Issue by scenaristeur** [novnc/noVNC#1778]
**Describe the bug**
HI, For a long time i was blocked, impossible to send the dot /period/NumpadDecimal on a french keyboard. 
the only way i've find is to change in the keyboard.js and add on line 50 . Instead of a keysym 46, i force it to 65454 (    XK_KP_Decimal:                  0xffae, in keysim.js ) otherwise i always get a semicolon `;`

```
  _sendKeyEvent(keysym, code, down) {
    if (down) {
      this._keyDownList[code] = keysym;
    } else {
      // Do we really think this key is down?
      if (!(code in this._keyDownList)) {
        return;
      }
      delete this._keyDownList[code];
    }
    if(keysym == 46){         <--- added line
      keysym = 65454}       <!---- added line
      Log.Debug("onkeyevent " + (down ? "down" : "up") +
      ", keysym: " + keysym, ", code: " + code);
      this.onkeyevent(keysym, code, down);
    }
```

Do you have an idea where is the problem and how i could have a dot/period/numpadDecimal without this ugly hack ? 

Thanxs

![Capture d’écran de 2023-05-11 12-50-48](https://github.com/novnc/noVNC/assets/4020744/bc3f64c3-e8e6-409b-b642-fd612918cecc)

My installation  : 
 - windows  10 in virtmanager with 
 - one Spice server and one VNC server 
 
![Capture d’écran de 2023-05-11 12-55-25](https://github.com/novnc/noVNC/assets/4020744/19bf0279-7316-4766-ba52-5f8a71db91b2)

![Capture d’écran de 2023-05-11 12-55-34](https://github.com/novnc/noVNC/assets/4020744/6b361055-52ab-460b-8f41-f09463a3df57)

(many 'clavier' / 'keyboard' on those screenshot , but problem is there even with only one default)

- then i run noVNC on 5912 port 
`noVNC$ ./utils/novnc_proxy --vnc localhost:5912`
- and websockify this way  `./websockify.py 5903 localhost:5902`
- i can now access my windows 10 vm in Firefox at http://host_ip:6080/vnc.html?host=host_ip&port=6080
- (host_ip is the ip address of the host)

The goal is to use Selenium to drive the Windows VM . so i have a Selenium / mocha that can access this firefox instance and drive it. 

using opencv lib and some tricky code , I can click on the désired zone in the Firefox window, all that works fine and i can connect to novnc, and launch Edge. 

then i use Selenium sendKey https://www.selenium.dev/documentation/webdriver/actions_api/keyboard/
 to send text or keybord action from that script and it works fine, and a AZERTY keyboard is well used

I've tried sendKeys(Key.DECIMAL) , alt +0x46 and so many, so many, so many other things but nothing works appart that hack mentionned above. 

and i 'have no choice other that splitting the url like this , otherwise, the dot/period/NumpadDecimal disappear 

```
    await driver.actions()
    .sendKeys('http:')
    .keyDown(Key.SHIFT)
    .sendKeys('::')
    .keyUp(Key.SHIFT)
    .sendKeys('novnc.')
    .sendKeys('com')
    .sendKeys(Key.ENTER)
    .perform()
```

**Client (please complete the following information):**
 - OS: windows10 vm hosted on Mint (  / Centos 7 running  the selenium script
 - Browser: firefox
 - Browser version: 91.4.1esr

**Server (please complete the following information):**
 - noVNC version: 1.4.0
 - VNC server: virt-manager 1.5.1 on Mint machine
 - WebSocket proxy: 0.11.0 novnc/websockify

**Comment by CendioOssman** [comment]
Unfortunately, it sounds like Selenium is not emulating a real keyboard properly and hence you are seeing issues. noVNC with QEMU will send physical key presses, which means that Selenium cannot take any shortcuts. If it wants to send a ".", it needs to send a combination of shift and ".".

I would suspect that you might not be able to use sendKeys() at all, unfortunately. Is Selenium even aware of keyboard layouts so it can emulate things correctly? Or does it just assume an American layout?

**Comment by scenaristeur** [comment]
hi @CendioOssman , thxs for your reply, but as you can see in the middle of the js console of this pic https://user-images.githubusercontent.com/4020744/237670529-bc3f64c3-e8e6-409b-b642-fd612918cecc.png 

a keydown event is received by novnc with key: '.' and keycode : "190" which mean "period" as i see here https://www.toptal.com/developers/keycode/period

and novnc util.js and keyboard.js (line 58) have interpreted them as "NumpadDecimal" so i think Selenium send a correct info and the issue is somewhere after that point

**Comment by CendioOssman** [comment]
Sorry, I thought you were trying to send the regular `.` key on the main section of the keyboard.

We don't generally use the keycode field as it is unreliable and only for legacy browsers. We also don't have 190 in the legacy mappings as far as I can tell.

The proper field is `code`, but it looks like Selenium is setting that properly, as the code manages to map things to `NumpadDecimal`. The fact that it doesn't pick the corresponding keysym indicates that the `location` field is set incorrectly. That is what determines which section of the keyboard is used, in the case where a symbol is present in multiple places.

**Comment by scenaristeur** [comment]
and according to this test          
`       expect(KeyboardUtil.getKeysym({code: 'NumpadDecimal', key: '.', location: 3})).to.be.equal(0xFFAE); ` on this line https://github.com/novnc/noVNC/blob/c2d6a06d6df7dcc5cda9567181c685cd1b84d750/tests/test.helper.js#L177

code: 'NumpadDecimal'  + key: '.'  should be 0xFFAE which is 65454, the keysym by the one i have to replace the given 46.

I am wrong ?

**Comment by scenaristeur** [comment]
> Sorry, I thought you were trying to send the regular `.` key on the main section of the keyboard.

I don't mind which one , the '.' on the main or the NumpadDecimal, I just want only one and i can't

**Comment by scenaristeur** [comment]
the issue seems to be around that line 

https://github.com/novnc/noVNC/blob/c2d6a06d6df7dcc5cda9567181c685cd1b84d750/core/input/domkeytable.js#L298

read by that one 

https://github.com/novnc/noVNC/blob/c2d6a06d6df7dcc5cda9567181c685cd1b84d750/core/input/util.js#L175

it seems that if i had a location == 3 , i would get `XK_KP_Decimal:                  0xffae,` (65454)

but i get `   XK_period:                      0x002e,`  (46 ) which seems to be location == 2

i must now find where that location is set

**Comment by CendioOssman** [comment]
That should be set by the browser. So if it is incorrect, it suggests that Selenium isn't setting it correctly.

Have you tried logging what Selenium sets the field to? Does it have a bad value? Or is it simply not set at all?

**Comment by CendioOssman** [comment]
Which browser are you using and this case? And on which platform? We'll need to see if we can reproduce it here and see where things go wrong.

**Comment by scenaristeur** [comment]
i'm on Ubuntu 18 with Firefox 102, where novnc / websockify are running and where i have made my last manual test. The selenium actor that i talk at the beginning is on centos 7 , FF 102 too,
i don't know if this can help, but on this page https://www.toptal.com/developers/keycode/numpad-decimal

when i use the decimal numpad AZERTY keyboard it is detected as 190  with a location 0 which is incorrect it shoulb be location 3 -> is it a firefox bug ?

![image](https://github.com/novnc/noVNC/assets/4020744/8895345c-5f76-429e-8b7e-d400c619f23a)

but on other computer it is detected as 110 

![image](https://github.com/novnc/noVNC/assets/4020744/3eb9dce4-15fd-4dd3-b8ea-5717f562ec0f)

both are numpadDecimal

**Comment by CendioOssman** [comment]
Yeah, something is unfortunately broken on your system. I'm not seeing that bug with Firefox 102 ESR on RHEL 8, or on RHEL 7.

Ubuntu 18.04 is EOL, so perhaps the issue is there. I would suggest an upgrade and see if that helps. Or check if you have some odd keyboard layout configured that is confusing Firefox.

**Comment by scenaristeur** [comment]
OK, sorry i will deep reinvestigate on my side. thousand of thxs for your help
